### PR TITLE
uxn: 0.pre+unstable=2021-08-30 -> unstable-2022-10-22

### PR DESCRIPTION
**Pull request by kototama**
###### Description of changes

Update Uxn to latest commit.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

It's my first contribution to NixOs, please show some understanding :green_heart:

### CODE REVIEW

**Comment by kototama** [comment]
@AndersonTorres can you have a look?

**Comment by SuperSandro2000** [comment]
darwin is broken

```
unpacking source archive /private/tmp/nix-build-source.drv-0/1b2049e238df96f32335edf1c6db35bd09f8b42d.tar.gz
building '/nix/store/sv60hk684hnl0nxsqarj6ssxvfymyhxh-uxn-unstable-2022-10-22.drv'...
unpacking sources
unpacking source archive /nix/store/lg0d6dmgkg0amzagxnj4wx20hpgnlbig-source
source root is source
patching sources
building
Cleaning..
./build.sh: line 84: brew: command not found
Building..
ld: warning: option -s is obsolete and being ignored
clang-11: error: no such file or directory: '/lib/libSDL2.a'
clang-11: error: no such file or directory: '/nix/store/znbmpsnibwsvi39i9mggjg1mrcmxjksr-SDL2-2.24.0/lib/libSDL2.a'
error: builder for '/nix/store/sv60hk684hnl0nxsqarj6ssxvfymyhxh-uxn-unstable-2022-10-22.drv' failed with exit code 1;
```

we must overwrite UXNEMU_LDFLAGS since we don't use brew.

**Comment by kototama** [comment]
> we must overwrite UXNEMU_LDFLAGS since we don't use brew.

I don't experience with Mac, how should it look like?

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -7369,6 +7369,12 @@
     githubId = 2037002;
     name = "Konstantinos";
   };
+  kototama = {
```
Comment (AndersonTorres) [inline maintainers/maintainer-list.nix:7372]: When editing this file to include yourself as a new maintainer, use a single commit to it (no need to do a separate pull request).

**Review by AndersonTorres — CHANGES_REQUESTED** [review]
I will drop my own PR, this one is cleaner.

File: pkgs/applications/emulators/uxn/default.nix [hunk 2]
```diff
@@ -6,13 +6,13 @@
 
 stdenv.mkDerivation rec {
```
Comment (AndersonTorres) [inline pkgs/applications/emulators/uxn/default.nix:7]:
```suggestion
stdenv.mkDerivation {
```

File: pkgs/applications/emulators/uxn/default.nix [hunk 3]
```diff
@@ -6,13 +6,13 @@
...
   pname = "uxn";
-  version = "0.pre+unstable=2021-08-30";
+  version = "unstable-2022-10-22";
 
   src = fetchFromSourcehut {
     owner = "~rabbits";
     repo = pname;
```
Comment (AndersonTorres) [inline pkgs/applications/emulators/uxn/default.nix:13]:
```suggestion
    repo = "uxn";
```

File: pkgs/applications/emulators/uxn/default.nix [hunk 4]
```diff
@@ -38,7 +34,7 @@ stdenv.mkDerivation rec {
 
     install -d $out/bin/ $out/share/${pname}/
```
Comment (AndersonTorres) [inline pkgs/applications/emulators/uxn/default.nix:35]:
```suggestion
    install -d $out/bin/ $out/share/uxn/
```

File: pkgs/applications/emulators/uxn/default.nix [hunk 5]
```diff
@@ -38,7 +34,7 @@ stdenv.mkDerivation rec {
...
 
-    cp uxnasm uxncli uxnemu $out/bin/
+    cp bin/uxnasm bin/uxncli bin/uxnemu $out/bin/
     cp -r projects $out/share/${pname}/
```
Comment (AndersonTorres) [inline pkgs/applications/emulators/uxn/default.nix:38]:
```suggestion
    cp -r projects $out/share/uxn/
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]
Do you know hot to make a unified diff?

File: maintainers/maintainer-list.nix (hunk 1 above)
Comment (kototama) [reply maintainers/maintainer-list.nix:7372]: Thanks, I have corrected that.

File: pkgs/applications/emulators/uxn/default.nix (hunk 4 above)
Comment (kototama) [reply pkgs/applications/emulators/uxn/default.nix:35]: what's th problem with using `pname` here?

File: pkgs/applications/emulators/uxn/default.nix (hunk 2 above)
Comment (kototama) [reply pkgs/applications/emulators/uxn/default.nix]: Ah I see why you remove pname now below. Why do you want to remove rec and the use of the pname variable?

Comment (AndersonTorres) [reply pkgs/applications/emulators/uxn/default.nix]: More recently, the use of `rec` is being avoided, because it triggers some undesirable behaviors like infinite recursions and bad interaction with overriding mechanisms.

In this specific case, just get rid of `pname` references around the code. It should suffice for now.

**Comment by kototama** [comment]
> darwin is broken

Ok I found a solution, let's see if it works with the CI...

File: pkgs/applications/emulators/uxn/default.nix [hunk 6]
```diff
@@ -21,14 +21,14 @@ stdenv.mkDerivation rec {
 
   dontConfigure = true;
 
-  # It is easier to emulate build.sh script
+  patchPhase = ''
```
Comment (AndersonTorres) [inline pkgs/applications/emulators/uxn/default.nix:24]:
```suggestion
  postPatch = ''
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]
This should suffice

File: pkgs/applications/emulators/uxn/default.nix (hunk 6 above)
Comment (kototama) [reply pkgs/applications/emulators/uxn/default.nix:24]: why?

Comment (AndersonTorres) [reply pkgs/applications/emulators/uxn/default.nix:24]: **Tradition**. Just follow it, _no questions_.

-------------------------------

Okay, I know this is not a satisfactory answer. Here is the best technical answer I can provide:

The most correct way to use custom `phase`s is

 ```nix
buildPhase = ''
  runHook preBuild

   # insert here the custom code you want

  runHook postBuild
'';
 ``` 

This`runHook` decorating the custom code is because Nixpkgs can be monkey-patched by the end user, via overlays, flakes etc., and these `pre/post` hooks help such _composability_. 

-----------------------

The `patchPhase` has another peculiarity: 

Usually we (and many other distros, for that matter) provide custom unified diff `.patch` files that modify the source code. These patch files are listed on the `patch = [ XXX.patch YYY.patch ... ];` attribute, and the default  `patchPhase` basically runs `patch -p1 < XXX.patch` on them.

When you rewrite the whole `patchPhase` like this

 ```nix
patchPhase = ''
  runHook prePatch

   # custom patch code, usually some crazy sed/awk one-liner

  runHook postPatch
'';
 ```

you leave no option to apply such `.patch`es, forcing the next developer to write the `patch -p1 < ...` code, effectively duplicating the effort of the default `patchPhase`.

That being said, the best when you want to use some sed-awk dark magic to modify the upstream code is to write it as a `postPatch`.

-------------------------

Now do it.

Comment (kototama) [reply pkgs/applications/emulators/uxn/default.nix:24]: Thanks for the explanations.
